jsc.exe --module-file should understand Windows paths

rkirsling opened the pull request:
#### 47fde25a21fbefda10d5052d36388bcbd45f32b6
<pre>
jsc.exe --module-file should understand Windows paths
<a href="https://bugs.webkit.org/show_bug.cgi?id=241518">https://bugs.webkit.org/show_bug.cgi?id=241518</a>

Reviewed by Yusuke Suzuki.

jsc.cpp&apos;s module loader was written without any accommodation for Windows, so:
1. On Windows, recognize C:\foo as an absolute path and .\foo and ..\foo as dotted relative paths (allowing &apos;/&apos; too).
2. On all platforms, stop misusing the URL(base, relative) constructor. This isn&apos;t the way to add file:/// to an abspath.

This ensures that module tests are able to run well on Windows.

* Source/JavaScriptCore/jsc.cpp:
(isAbsolutePath): Added.
(isDottedRelativePath): Added.
(absoluteFileURL): Renamed from `absolutePath`.
(GlobalObject::moduleLoaderImportModule):
(GlobalObject::moduleLoaderResolve):
(JSC_DEFINE_HOST_FUNCTION):
(computeFilePath):
(runWithOptions):

Canonical link: <a href="https://commits.webkit.org/251514@main">https://commits.webkit.org/251514@main</a>
git-svn-id: <a href="https://svn.webkit.org/repository/webkit/trunk@295509">https://svn.webkit.org/repository/webkit/trunk@295509</a> 268f45cc-cd09-0410-ab3c-d52691b4dbfc
</pre>

Constellation commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -844,8 +844,37 @@ static URL currentWorkingDirectory()
+    // Just look for local drives like C:\.
+    return path.length() > 2 && isASCIIAlpha(path[0]) && path[1] == ':' && (path[2] == '\\' || path[2] == '/');
+#else
+    return path.startsWith('/');
```
Since it is used for module specifier too, I think / root path should be supported too in Windows. And in that case, we should make a path from the current path's drive.

rkirsling replied:
Hmm, are you sure? It seems like other engines still treat that as a relative path on Windows:

test.js
```
import result from '/GitHub/WebKit/test-module.js';
print(result);
```

Result:
```
PS C:\GitHub\WebKit> ~\.jsvu\v8.cmd --module test.js
undefined:0: Error: d8: Error reading  module from C:/GitHub/WebKit//GitHub/WebKit/test-module.js
    imported by C:/GitHub/WebKit/test.js

PS C:\GitHub\WebKit> ~\.jsvu\sm.cmd --module test.js
Error: can't open C:\GitHub\WebKit\\GitHub\WebKit\test-module.js: No such file or directory
Stack:
```

Constellation reviewed (approved):
r=me

rkirsling committed:
jsc.exe --module-file should understand Windows paths
https://bugs.webkit.org/show_bug.cgi?id=241518

Reviewed by Yusuke Suzuki.

jsc.cpp's module loader was written without any accommodation for Windows, so:
1. On Windows, recognize C:\foo as an absolute path and .\foo and ..\foo as dotted relative paths (allowing '/' too).
2. On all platforms, stop misusing the URL(base, relative) constructor. This isn't the way to add file:/// to an abspath.

This ensures that module tests are able to run well on Windows.

* Source/JavaScriptCore/jsc.cpp:
(isAbsolutePath): Added.
(isDottedRelativePath): Added.
(absoluteFileURL): Renamed from `absolutePath`.
(GlobalObject::moduleLoaderImportModule):
(GlobalObject::moduleLoaderResolve):
(JSC_DEFINE_HOST_FUNCTION):
(computeFilePath):
(runWithOptions):

Canonical link: https://commits.webkit.org/251514@main
git-svn-id: https://svn.webkit.org/repository/webkit/trunk@295509 268f45cc-cd09-0410-ab3c-d52691b4dbfc
